HW2/P3/mandelbrot.pyx

```diff
@@ -16,62 +22,45 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                  int max_iterations=511):
     cdef:
        int i, j, iter
-       np.complex64_t c, z
-
-       # To declare AVX.float8 variables, use:
-       # cdef:
-       #     AVX.float8 v1, v2, v3
-       #
-       # And then, for example, to multiply them
-       #     v3 = AVX.mul(v1, v2)
-       #
-       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+       AVX.float8 c_real, c_imag, z_real, z_imag, z_temp, iters, curr
 
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
 
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        # multithreading over all rows
+        for i in prange(in_coords.shape[0], num_threads=4, schedule='static', chunksize=1):
+            # looping through every 8 columns (instruction-level parallelism)
+            for j in xrange(0,in_coords.shape[1],8):
+                # split c into real and imaginary parts
+                c_real = AVX.make_float8((in_coords[i, j+7]).real,(in_coords[i, j+6]).real,(in_coords[i, j+5]).real,(in_coords[i, j+4]).real,(in_coords[i, j+3]).real,(in_coords[i, j+2]).real,(in_coords[i, j+1]).real,(in_coords[i, j]).real)
+                c_imag = AVX.make_float8((in_coords[i, j+7]).imag,(in_coords[i, j+6]).imag,(in_coords[i, j+5]).imag,(in_coords[i, j+4]).imag,(in_coords[i, j+3]).imag,(in_coords[i, j+2]).imag,(in_coords[i, j+1]).imag,(in_coords[i, j]).imag)
+
+                # make z into a float8 vector with value 0 across
+                z_real = AVX.float_to_float8(0.0)
+                z_imag = AVX.float_to_float8(0.0)
+
+                # initialize AVX iterate counts to 0.0 and curr to 0.0
+                iters = AVX.float_to_float8(0.0)
+                		
+                # iterate until loop is broken or max_iterations is reached
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+                    # create AVX register containing which elements have |z|^2 < 4
+                    curr = AVX.less_than(AVX.add(AVX.mul(z_real, z_real),AVX.mul(z_imag, z_imag)), AVX.float_to_float8(4.0))
```

Comment:
I didn't know you could call AVX functions without saving to a local variable, this looks really clean.